### ISSUE TITLE
hpack is slow due to recursion into non-build directories

### ISSUE THREAD

**Issue by jdreaver** [sol/hpack#165]
Hello! We have a fairly large haskell project that uses `hpack` via `stack`. We noticed it was taking `hpack` up to 5 seconds to turn a `package.yaml` file into a cabal file, even if nothing changed. I ran `strace hpack` and discovered that:
* `hpack` was recursing into the `.stack-work` directory, which just so happened to be huge on my machine because I was testing different resolvers.
* `hpack` was also recursing into a vendored mathjax directory, which has tons of files but is not mentioned in `package.yaml`. It just happens to be in the same directory as the haskell project.

Is it possible to tell `hpack` to ignore certain directories? Should it ignore `.stack-work` by default? Actually, should it ignore any directory that is not mentioned in `package.yaml`? We specify `source-dirs` for our library and our executable, so I feel like it should be able to contain itself to just those directories when searching for Haskell files. We could move the mathjax stuff, but wasting time reading all of our `.stack-work` directories is a pretty big price to pay every time we run a `stack` command.

Thanks in advance for any help!

**Comment by sol** [comment]
@jdreaver Hey, this sounds like something we want to fix fast!

Right now I have know idea where this behavior comes from, nor whether it's `hpack` itself or some of it's dependencies (e.g. `Glob`).

Can you help to try narrowing it down?  Specifically, let's try to produce a minimal example that exposes the issue.

Can you try some minimal example that check if it exposes the issue?  Something like this:
```yaml
name: foo
library:
  source-dirs: src
  exposed-modules: Foo
  other-modules: Bar
```

**Comment by jdreaver** [comment]
You know what, I just discovered that we had this as the `library` block:

```yaml
library:
  source-dirs: .
```

Obviously if you do that then `hpack` has reason to believe any directory under the current directory could contain Haskell modules. Using `source-dirs: .` is a somewhat common pattern with small scripts though, should `hpack` make a special exception for `.stack-work` in the current directory?

Anyway I'm gonna chalk this up to user error... I think I had too many similar-looking `package.yaml` files open when I first found the issue :sweat_smile: Sorry for the noise!

**Comment by mitchellwrosen** [comment]
I could be wrong, but don't haskell module names have to mirror the directory tree exactly, so no module will be found in any folder that doesn't start with a capital letter?

**Comment by jdreaver** [comment]
> I could be wrong, but don't haskell module names have to mirror the directory tree exactly, so no module will be found in any folder that doesn't start with a capital letter?

That's a good point. `hpack` could only recurse into directories starting with capital letters starting at all `source-dirs`.

**Comment by sol** [comment]
After we classified this as a user error, I didn't consider this crucial anymore.  However, it's fixed now and on Hackage as `0.17.1`.

**Comment by jdreaver** [comment]
Great! That looks like a solid change.
